vermouth/gmx/itp_read.py

```diff
@@ -244,289 +208,187 @@ def _block(self, line, lineno=0):
-        parameters.append(parameter)
-    return parameters
-
-
-def _is_param_effector(token):
-    return (
-        '(' in token
-        and not token.startswith('(')
-        and token.endswith(')')
-    )
-
-
-def _base_parser(tokens, context, context_type, section, current_meta, natoms=None, delete=False):
-
-    # Group the atoms and their attributes
-    atoms = _get_atoms(tokens, natoms)
-
-    if natoms is not None and len(atoms) != natoms:
-        raise IOError('Found {} atoms while {} were expected.'
-                      .format(len(atoms), natoms))
-
-    # Normalize the atom references.
-    # Blocks and links treat these references differently.
-    # For blocks:
-    # * references can be written as indices or atom names
-    # * a reference cannot be prefixed by + or -
-    # * an interaction cannot create a new atom
-    # For links:
-    # * references must be atom names, but they can be prefixed with one or
-    #   more + or - to signify the order in the sequence
-    # * interactions create nodes
-
-    treated_atoms = _treat_block_interaction_atoms(atoms, context, section)
-
-    # Everything that is not atoms are the interaction parameters    
-    parameters = _parse_interaction_parameters(tokens)
-
-    apply_to_all_interactions = context._apply_to_all_interactions[section]
-
-    if current_meta:
-       meta = {current_meta['condition']:current_meta['tag']}
-    else:
-       meta = {} #dict(collections.ChainMap(meta, apply_to_all_interactions))
- 
-    interaction = Interaction(
-                  atoms=treated_atoms,
-                  parameters=parameters,
-                  meta=meta,
-                  )
-
-    interaction_list = context.interactions.get(section, [])
-    interaction_list.append(interaction)
-    context.interactions[section] = interaction_list
-
-
-def _parse_block_atom(tokens, context):
-
-    # deque does not support slicing
-    first_six = (tokens.popleft() for _ in range(6))
-    index, atype, resid, resname, name, charge_group = first_six
-    
-    if str(index) in context:
-        msg = ('There is already an atom named "{}" in the block "{}". '
-               'Atom names must be unique within a block.')
-        raise IOError(msg.format(name, context.name))
-
-    atom = {
-        'atomname': name,
-        'atype': atype,
-        'resname': resname,
-        'resid': int(resid),
-        'charge_group': int(charge_group),
-    }
-
-    # charge and mass are optional, but charge has to be defined for mass to be
-    if tokens:
-        atom['charge'] = float(tokens.popleft())
-    if tokens:
-        atom['mass'] = float(tokens.popleft())
-
-    attributes={}
-    context.add_atom_from_index(dict(collections.ChainMap(attributes, atom)), index=index)
+            atom['charge'] = float(tokens.popleft())
+        if tokens:
+            atom['mass'] = float(tokens.popleft())
+
+        attributes = {}
+        context.add_node(index, **dict(collections.ChainMap(attributes, atom)))
```

Comment:
No need to ChainMap if attributes is {}. Just `context.add_node(index, atom)` will do. But see above, I would generate a new number here, and use that as node index, rather than index directly.

Reply:
@pckroon not sure I understand this? So attributes needs to be a dict of what? If I simply plug in your suggestion I get an attribute error and I believe throughout the rest of the code the chain mapping is also used in these kinds of situations.